lib/utils/compose.js

```diff
@@ -0,0 +1,1086 @@
+					// before creating a date out of them, as `new Date(null)`
+					// creates a date representing UNIX time 0.
+					if (builtImage.startTime) {
+						image.props.startTime = new Date(builtImage.startTime);
+					}
+					if (builtImage.endTime) {
+						image.props.endTime = new Date(builtImage.endTime);
+					}
+					return docker
+						.getImage(image.name)
+						.inspect()
+						.get('Size')
+						.then(size => {
+							image.props.size = size;
+						})
+						.return(image);
+				},
+			).tap(function(images) {
+				const summary = _(images)
+					.map(({ serviceName, props }) => [
+						serviceName,
+						`Image size: ${humanize.filesize(props.size)}`,
+					])
+					.fromPairs()
+					.value();
+				renderer.end(summary);
+			});
+		})
+		.finally(renderer.end);
+}
+
+const createRelease = function(apiEndpoint, auth, userId, appId, composition) {
+	const _ = require('lodash');
+	const crypto = require('crypto');
+	const releaseMod = require('balena-release');
+
+	const client = releaseMod.createClient({ apiEndpoint, auth });
+
+	return releaseMod
+		.create({
+			client,
+			user: userId,
+			application: appId,
+			composition,
+			source: 'local',
+			commit: crypto
+				.pseudoRandomBytes(16)
+				.toString('hex')
+				.toLowerCase(),
+		})
+		.then(function({ release, serviceImages }) {
+			return {
+				client,
+				release: _.omit(release, [
+					'created_at',
+					'belongs_to__application',
+					'is_created_by__user',
+					'__metadata',
+				]),
+				serviceImages: _.mapValues(serviceImages, serviceImage =>
+					_.omit(serviceImage, [
+						'created_at',
+						'is_a_build_of__service',
+						'__metadata',
+					]),
+				),
+			};
+		});
+};
+
+const tagServiceImages = (docker, images, serviceImages) =>
+	Promise.map(images, function(d) {
+		const serviceImage = serviceImages[d.serviceName];
+		const imageName = serviceImage.is_stored_at__image_location;
+		const match = /(.*?)\/(.*?)(?::([^/]*))?$/.exec(imageName);
+		if (match == null) {
+			throw new Error(`Could not parse imageName: '${imageName}'`);
+		}
+		const [, registry, repo, tag = 'latest'] = match;
+		const name = `${registry}/${repo}`;
+		return docker
+			.getImage(d.name)
+			.tag({ repo: name, tag, force: true })
+			.then(() => docker.getImage(`${name}:${tag}`))
+			.then(localImage => ({
+				serviceName: d.serviceName,
+				serviceImage,
+				localImage,
+				registry,
+				repo,
+				logs: d.logs,
+				props: d.props,
+			}));
+	});
+
+const getPreviousRepos = (sdk, docker, logger, appID) =>
+	sdk.pine
+		.get({
+			resource: 'release',
+			options: {
+				$filter: {
+					belongs_to__application: appID,
+					status: 'success',
+				},
+				$select: ['id'],
+				$expand: {
+					contains__image: {
+						$expand: 'image',
+					},
+				},
+				$orderby: 'id desc',
+				$top: 1,
+			},
+		})
+		.then(function(release) {
+			// grab all images from the latest release, return all image locations in the registry
+			if (release.length > 0) {
```

Comment:
This used to be
```coffee
release?.length > 0
```
which is a bad comparison, the compilation is
```js
(release != null ? release.length : undefined) > 0
```
I changed it to treat it as existing because pinejs-client should always return an array here anyway